h2/src/main/org/h2/expression/BinaryOperation.java

```diff
@@ -165,14 +179,18 @@ public Expression optimize(Session session) {
         switch (opType) {
         case CONCAT: {
             TypeInfo l = left.getType(), r = right.getType();
-            if (DataType.isStringType(l.getValueType()) && DataType.isStringType(r.getValueType())) {
-                long precision = l.getPrecision() + r.getPrecision();
-                if (precision >= 0 && precision < Integer.MAX_VALUE) {
-                    type = TypeInfo.getTypeInfo(Value.STRING, precision, 0, null);
-                    break;
+            if (l.getValueType() == Value.ARRAY) {
```

Comment:
should this not be 
    if (l.getValueType() == Value.ARRAY || .rgetValueType() == Value.ARRAY)) 
?

So  that
    SELECT 3 || ARRAY[1, 2] 
produces
   [1, 2, 3]
?

Reply:
Definition is not very strict:
```
<array concatenation> ::=
    <array value expression 1> <concatenation operator> <array primary>

<array value expression 1> ::= <array value expression>

<array value expression> ::= <array concatenation> | <array primary>

<array primary> ::= <array value function> | <value expression primary>

1) The declared type of the <array value expression> is the declared type of the
immediately contained <array concatenation> or <array primary>.

2) The declared type of <array primary> is the declared type of the immediately
contained <array value function> or <value expression primary>, which shall be
an array type or a distinct type whose source type is an array type.
```
So technically both sides shall be arrays.

However, PostgreSQL treats concatenation operator as a part of array concatenation when only the right side is an array.

Reply:
Because string concatenation with one array operand is meaningless (actually its both sides described as character strings in the standard), let's follow the PosgreSQL.

Reply:
a434a6a2b3e261760c42b2edbaa1334e04a60d5d